xla/service/algebraic_simplifier.cc

```diff
@@ -736,6 +736,16 @@ bool AlgebraicSimplifierVisitor::ReplaceInstructionIfCompatible(
 bool AlgebraicSimplifierVisitor::ReplaceInstructionIfCompatible(
     HloInstruction* old_instruction,
     absl::Span<HloInstruction* const> new_instructions) {
+  // It's tricky for the simplifier to determine whether
+  // it should remove the op when control deps are present. I.e.
+  // control deps might be added to preserve a certain order.
+  // It's better to not process in that case.
```

Comment:
Would it make sense to rewire such control dependencies? If the op is redundant and all we want is the transitively implied control dependencies, can we preserve those and remove the op?

Reply:
It's difficult to determine the function of the control dependency, if it's added to preserve a certain order, then we probably can't treat it a a redundant op, it has a purpose in the graph. I wanted to avoid changing the dependency chain too much.
An example will be:
a = op(b)
e=op(x)
f=op(y)
c = op(a), control_predecessor={e,f}
h=op(a)
let's say a and c are qualified to be fused.
Without rewiring, a and h could be executed before e and f, but if we rewire the control deps, 
e=op(x)
f=op(y)
a = op(b), control_predecessor={e,f}
h=op(a)
a and h now must be executed after e and f.
If h is a collective, the schedule will be sub-optimal.

Reply:
Makes sense. Thanks for the example.